echopop/nwfsc_feat/apportion.py

```diff
@@ -1,123 +1,704 @@
+        Dictionary mapping column names from mesh_data_df to biodata column names for joining.
+        Example: {"geostratum_ks": "stratum_ks"}
+    stratum_weights_df : pd.DataFrame
+        DataFrame containing average weights per stratum for converting biomass to abundance.
+        Index should match the linked columns from mesh_biodata_link.
+    stratum_sigma_bs_df : pd.DataFrame
+        DataFrame containing sigma_bs (backscattering coefficient) values per stratum.
+        Must have 'sigma_bs' column with index matching linked columns.
+    group_by : List[str], default []
+        List of column names to group biological data by (e.g., ["sex", "age_bin"]).
+
+    Returns
+    -------
+    None
+        Function modifies mesh_data_df in place, adding new columns for biomass, abundance,
+        and NASC values.
+
+    Raises
+    ------
+    KeyError
+        If required link columns are missing from biodata or mesh DataFrames.
+
+    Notes
+    -----
+    The function performs the following steps:
+    1. Computes biomass from biomass_density x area if not present
+    2. Links mesh data to biological proportions using mesh_biodata_link
+    3. Applies proportions to distribute total biomass across biological groups
+    4. Converts biomass to abundance using stratum weights
+    5. Calculates NASC using abundance and sigma_bs values
+
+    The NASC calculation uses the formula: NASC = abundance x sigma_bs x 4π
+    """
+
+    # Convert to a Dictionary if needed
+    if isinstance(biodata, pd.DataFrame):
+        biodata = {"": biodata}
+
+    # Get the column indices that will be used for setting shared indices
+    column_names = list(
+        set(
+            col
+            for df in biodata.values()
+            if isinstance(df.columns, pd.Index)
+            for col in df.columns.names
+        )
+    )
+
+    # Check
+    # ---- Setdiff: biodata
+    missing_biodata_columns = set(list(mesh_biodata_link.values())).difference(column_names)
+    if len(missing_biodata_columns) > 0:
+        raise KeyError(
+            f"The following link columns were missing from the biological data: "
+            f"{', '.join(col for col in missing_biodata_columns)}."
+        )
+    # ---- Setdiff: mesh
+    missing_mesh_columns = set(list(mesh_biodata_link.keys())).difference(mesh_data_df.columns)
+    if len(missing_mesh_columns) > 0:
+        raise KeyError(
+            f"The following link columns were missing from the mesh DataFrame: "
+            f"{', '.join(col for col in missing_mesh_columns)}."
+        )
+
+    # Compute biomass if not already in DataFrame
+    if "biomass" not in mesh_data_df.columns:
+        mesh_data_df["biomass"] = mesh_data_df["biomass_density"] * mesh_data_df["area"]
+
+    # Set the link column names
+    mesh_data_df.rename(columns=mesh_biodata_link, inplace=True)
+
+    # Map the link columns
+    mesh_index = [v for v in mesh_biodata_link.values() if v in mesh_data_df.columns]
```

Comment:
If you just renamed the column, why would the renamed column not exist in the dataframe and needed to be checked?